File: LeetCode/Top-Interview-150/HashMap/HappyNumber.py
class Solution:
    
    def isHappy(self, n: int) -> bool:

        # Function to calculate the Sum of Squares

        def SumofSquares(num):

            sum = 0

            while num:

                digit = num % 10
                sum += digit ** 2
                num //= 10
            
            return sum

        # HashMap to keep track of sums

        sums = {}

        while n!=1 and n not in sums:

            sums[n] = True # Marking the current sum as "tracked"
            n = SumofSquares(n) # Calculate the next

        return n == 1 # Returned if it's a happy number
    
'''

Time Complexity:

- O(log n)

'''

File: LeetCode/Top-Interview-150/HashMap/HappyNumber.java
public class Solution 
{
    
    // Helper method to calculate the sum of squares of digits of a number
    private int sumOfSquares(int num) 
    {
        int sum = 0;

        while (num > 0) 
        {
            int digit = num % 10;
            sum += digit * digit;
            num /= 10;
        }

        return sum;
    }

    public boolean isHappy(int n) 
    {
        // HashMap to keep track of sums we have seen
        HashMap<Integer, Boolean> sums = new HashMap<>();
        
        while (n != 1 && !sums.containsKey(n)) 
        {
            sums.put(n, true); // Mark the current sum as seen
            n = sumOfSquares(n); // Calculate the next sum of squares
        }
        
        // If n is 1, it's a happy number; otherwise, it's not
        return n == 1;
    }
}

/*

Time Complexity:

- O(log n)

*/